govcd/system.go

```diff
@@ -528,3 +587,111 @@ func QueryOrgVdcNetworkByName(vcdCli *VCDClient, name string) ([]*types.QueryRes
+	if err != nil {
+		return nil, err
+	}
+
+	return org, nil
+}
+
+// GetOrgById finds an Organization by ID
+// On success, returns a pointer to the Org structure and a nil error
+// On failure, returns a nil pointer and an error
+func (vcdClient *VCDClient) GetOrgById(orgId string) (*Org, error) {
+	orgUrl, err := getOrgHREFById(vcdClient, orgId)
+	if err != nil {
+		// Since this operation is a lookup from a list, we return the standard ErrorEntityNotFound
+		return nil, ErrorEntityNotFound
+	}
+	org := NewOrg(&vcdClient.Client)
+
+	_, err = vcdClient.Client.ExecuteRequest(orgUrl, http.MethodGet,
+		"", "error retrieving org list: %s", nil, org.Org)
+	if err != nil {
+		return nil, err
+	}
+
+	return org, nil
+}
+
+// GetOrgByNameOrId finds an Organization by name or ID
+// On success, returns a pointer to the Org structure and a nil error
+// On failure, returns a nil pointer and an error
+func (vcdClient *VCDClient) GetOrgByNameOrId(identifier string) (*Org, error) {
+
+	org, err := vcdClient.GetOrgById(identifier)
+	if err != nil {
+		org, err = vcdClient.GetOrgByName(identifier)
+	}
+	return org, err
+}
+
+// GetAdminOrgByName finds an Admin Organization by name
+// On success, returns a pointer to the Admin Org structure and a nil error
+// On failure, returns a nil pointer and an error
+func (vcdClient *VCDClient) GetAdminOrgByName(orgName string) (*AdminOrg, error) {
+	orgUrl, err := getOrgHREF(vcdClient, orgName)
+	if err != nil {
+		return nil, ErrorEntityNotFound
+	}
+	orgHREF := vcdClient.Client.VCDHREF
+	orgHREF.Path += "/admin/org/" + strings.Split(orgUrl, "/api/org/")[1]
+
+	adminOrg := NewAdminOrg(&vcdClient.Client)
+
+	_, err = vcdClient.Client.ExecuteRequest(orgHREF.String(), http.MethodGet,
+		"", "error retrieving org: %s", nil, adminOrg.AdminOrg)
+	if err != nil {
+		return nil, err
+	}
+
+	return adminOrg, nil
+}
+
+// GetAdminOrgById finds an Admin Organization by ID
+// On success, returns a pointer to the Admin Org structure and a nil error
+// On failure, returns a nil pointer and an error
+func (vcdClient *VCDClient) GetAdminOrgById(orgId string) (*AdminOrg, error) {
+	orgUrl, err := getOrgHREFById(vcdClient, orgId)
+	if err != nil {
+		return nil, ErrorEntityNotFound
+	}
+	orgHREF := vcdClient.Client.VCDHREF
+	orgHREF.Path += "/admin/org/" + strings.Split(orgUrl, "/api/org/")[1]
+
+	adminOrg := NewAdminOrg(&vcdClient.Client)
+
+	_, err = vcdClient.Client.ExecuteRequest(orgHREF.String(), http.MethodGet,
+		"", "error retrieving org: %s", nil, adminOrg.AdminOrg)
+	if err != nil {
+		return nil, err
+	}
+
+	return adminOrg, nil
+}
+
+// GetAdminOrgByNameOrId finds an Admin Organization by name or ID
+// On success, returns a pointer to the Admin Org structure and a nil error
+// On failure, returns a nil pointer and an error
+func (vcdClient *VCDClient) GetAdminOrgByNameOrId(identifier string) (*AdminOrg, error) {
+
+	adminOrg, err := vcdClient.GetAdminOrgById(identifier)
+	if err != nil {
```

Comment:
here if error isn't `entity not foun` then it's lost

Reply:
We don't know if `identifier` contains a name or an ID.
The most probable reason that this call fails is that we are looking by ID using a name. It is expected that it fails in this case, and we immediately search by name.
Whatever is returned at that point, is a legitimate error.

Reply:
Changed the implementation of `GetAdminOrgByNameOrId` (and also `GetOrgByNameOrId` and `GetUserByNameOrId`) to make it more readable and robust